File: serialization/serializers/commons/jackson/src/main/java/com/cedarsoft/serialization/serializers/jackson/UUIDSerializer.java
package com.cedarsoft.serialization.serializers.jackson;

import java.io.IOException;
import java.util.UUID;

import javax.annotation.Nonnull;
import javax.inject.Inject;

import com.cedarsoft.serialization.jackson.AbstractJacksonSerializer;
import com.cedarsoft.serialization.jackson.StringSerializer;
import com.cedarsoft.version.Version;
import com.cedarsoft.version.VersionException;
import com.cedarsoft.version.VersionRange;
import com.fasterxml.jackson.core.JsonGenerator;
import com.fasterxml.jackson.core.JsonParser;
import com.fasterxml.jackson.core.JsonProcessingException;

/**
 * @author Johannes Schneider (<a href="mailto:js@cedarsoft.com">js@cedarsoft.com</a>)
 */
public class UUIDSerializer extends AbstractJacksonSerializer<UUID> {
  @Inject
  public UUIDSerializer() {
    super("uuid", VersionRange.from(1, 0, 0).to());
  }

  @Override
  public boolean isObjectType() {
    return false;
  }

  @Override
  public void serialize(@Nonnull JsonGenerator serializeTo, @Nonnull UUID object, @Nonnull Version formatVersion) throws IOException, VersionException, JsonProcessingException {
    verifyVersionWritable(formatVersion);
    serializeTo.writeString(object.toString());
  }

  @Override
  @Nonnull
  public UUID deserialize(@Nonnull JsonParser deserializeFrom, @Nonnull Version formatVersion) throws IOException, VersionException, JsonProcessingException {
    verifyVersionReadable(formatVersion);
    String uuidAsString = deserializeFrom.getText().trim();

    return UUID.fromString(uuidAsString);
  }
}

File: commons/kotlinx-serializers/src/main/java/com/cedarsoft/commons/kotlin/serializers/UUIDSerializer.kt
package com.cedarsoft.commons.kotlin.serializers

import kotlinx.serialization.Decoder
import kotlinx.serialization.Encoder
import kotlinx.serialization.KSerializer
import kotlinx.serialization.PrimitiveDescriptor
import kotlinx.serialization.PrimitiveKind
import kotlinx.serialization.SerialDescriptor
import kotlinx.serialization.Serializer
import java.util.UUID

/**
 * Serializer for UUID
 * @author Johannes Schneider ([js@cedarsoft.com](mailto:js@cedarsoft.com))
 */
@Serializer(forClass = UUID::class)
object UUIDSerializer : KSerializer<UUID> {
  override val descriptor: SerialDescriptor = PrimitiveDescriptor("UUID", PrimitiveKind.STRING)

  override fun serialize(encoder: Encoder, obj: UUID) {
    encoder.encodeString(obj.toString())
  }

  override fun deserialize(decoder: Decoder): UUID {
    return UUID.fromString(decoder.decodeString())
  }
}
